### ISSUE TITLE
Wrong coefficients in CyclotomicField(6,z)

### ISSUE THREAD

**Issue by cheque** [Nemocas/Nemo.jl#1039]
In the cyclotomic field of order 6, polynomials of degree 4 and 5 appear to be returning trash as coefficients...
It appears to be the only case, but it is worth to check it out.

This code:

	Collect_res = Matrix{Float64}[]
	for m = 1:10
		Cm, z = CyclotomicField(m, "z")
		mat = zeros((m,m));
		for i = 1:m
			for j= 1:m
				mat[i,j] = float(Rational(coeff(z^(i-1),j-1)))
			end
		end
		push!(Collect_res,mat)
	end

Produces this result:
`
  Matrix{Float64}
  
  1×1 Matrix{Float64}:
   1.0
  
  2×2 Matrix{Float64}:
    1.0   1.0
   -1.0  -1.0
  
  3×3 Matrix{Float64}:
    1.0   0.0  0.0
    0.0   1.0  0.0
   -1.0  -1.0  0.0
  
  4×4 Matrix{Float64}:
    1.0   0.0  0.0  1.0
    0.0   1.0  0.0  1.0
   -1.0   0.0  0.0  1.0
    0.0  -1.0  0.0  1.0
  
  5×5 Matrix{Float64}:
    1.0   0.0   0.0   0.0  0.0
    0.0   1.0   0.0   0.0  0.0
    0.0   0.0   1.0   0.0  0.0
    0.0   0.0   0.0   1.0  0.0
   -1.0  -1.0  -1.0  -1.0  0.0
  
  **6×6 Matrix{Float64}:
    1.0   0.0  0.0  1.0  1.40306e14  1.40306e14
    0.0   1.0  0.0  1.0  1.40306e14  1.40308e14
   -1.0   1.0  0.0  1.0  1.40306e14  1.40306e14
   -1.0   0.0  0.0  1.0  1.40306e14  1.40306e14
    0.0  -1.0  0.0  1.0  1.40306e14  1.40306e14
    1.0  -1.0  0.0  1.0  1.40306e14  1.40306e14**
  
  7×7 Matrix{Float64}:
    1.0   0.0   0.0   0.0   0.0   0.0  0.0
    0.0   1.0   0.0   0.0   0.0   0.0  0.0
    0.0   0.0   1.0   0.0   0.0   0.0  0.0
    0.0   0.0   0.0   1.0   0.0   0.0  0.0
    0.0   0.0   0.0   0.0   1.0   0.0  0.0
    0.0   0.0   0.0   0.0   0.0   1.0  0.0
   -1.0  -1.0  -1.0  -1.0  -1.0  -1.0  0.0
  
  8×8 Matrix{Float64}:
    1.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0
    0.0   1.0   0.0   0.0  0.0  0.0  0.0  0.0
    0.0   0.0   1.0   0.0  0.0  0.0  0.0  0.0
    0.0   0.0   0.0   1.0  0.0  0.0  0.0  0.0
   -1.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0
    0.0  -1.0   0.0   0.0  0.0  0.0  0.0  0.0
    0.0   0.0  -1.0   0.0  0.0  0.0  0.0  0.0
    0.0   0.0   0.0  -1.0  0.0  0.0  0.0  0.0
  
  9×9 Matrix{Float64}:
    1.0   0.0   0.0   0.0   0.0   0.0  0.0  0.0  0.0
    0.0   1.0   0.0   0.0   0.0   0.0  0.0  0.0  0.0
    0.0   0.0   1.0   0.0   0.0   0.0  0.0  0.0  0.0
    0.0   0.0   0.0   1.0   0.0   0.0  0.0  0.0  0.0
    0.0   0.0   0.0   0.0   1.0   0.0  0.0  0.0  0.0
    0.0   0.0   0.0   0.0   0.0   1.0  0.0  0.0  0.0
   -1.0   0.0   0.0  -1.0   0.0   0.0  0.0  0.0  0.0
    0.0  -1.0   0.0   0.0  -1.0   0.0  0.0  0.0  0.0
    0.0   0.0  -1.0   0.0   0.0  -1.0  0.0  0.0  0.0
  
  10×10 Matrix{Float64}:
    1.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0   1.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0   0.0   1.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0   0.0   0.0   1.0  0.0  0.0  0.0  0.0  0.0  0.0
   -1.0   1.0  -1.0   1.0  0.0  0.0  0.0  0.0  0.0  0.0
   -1.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0  -1.0   0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0   0.0  -1.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
    0.0   0.0   0.0  -1.0  0.0  0.0  0.0  0.0  0.0  0.0
    1.0  -1.0   1.0  -1.0  0.0  0.0  0.0  0.0  0.0  0.0
  ]

`

**Comment by thofma** [comment]
The field has degree 2, so coefficients of degree 4 and 5 make no sense (the function should probably error).

**Comment by cheque** [comment]
It should not matter the degree of the field, it should return zero, as in the other cases.

**Comment by thofma** [comment]
I agree, the other cases are also broken.

Do you have a preference @wbhart? The non-quadratic cases are probably working because `fmpq_poly_get_coeff` is doing the right thing. Should we just disallow indices beyong the field degree?

**Comment by wbhart** [comment]
Sounds ok to me. Hopefully we are not relying on that anywhere to access unreduced polynomials or something like that.

To be honest I didn't even know we had a coeff function for number field elements. I guess that is what needs fixing in Nemo?

**Comment by cheque** [comment]
I would not suggest disallowing indices beyond the field degree...
There are some computations where one will obtain a higher degree polynomial and it would be OK to have this built-on reduction working. I think disallowing indices beyond the field degree would be a disservice to the whole thing...
It makes perfect sense to make some computations and rely on the correct reduction via the library functions.

**Comment by thofma** [comment]
Do you have an opinion on this @fieker?

**Comment by fieker** [comment]
it needs fixing in antic. We can fix it in Nemo instead, but the error is in antic. fmpq_poly_get_coeff_fmpq fills with zero, 
fmpz_set(fmpq_numref(a), bnum + i) destroyes the memory if i is too large

**Comment by wbhart** [comment]
I've fixed the issue in antic, but the issue reported by OP must be in Flint, as the degree there is more than 2.

I've made it so that if you read beyond the length of an *unreduced* element then it returns 0 (otherwise it doesn't seem very useful). The set_coeff_num_fmpz function now raises an exception if you try to write beyond this point.

If you stick to functions that do reduction (which is most of them) then you won't have any problems with unreduced vs reduced. If you don't, you are on your own.
